pymc/backends/zarr.py

```diff
@@ -0,0 +1,836 @@
+
+        This function creates and fills with default values the groups below the
+        ``ZarrTrace.root`` group. It creates the ``constant_data``, ``observed_data``,
+        ``posterior``, ``unconstrained_posterior`` (if ``include_transformed = True``),
+        ``sample_stats``, and ``_sampling_state`` zarr groups, and all of the relevant
+        arrays that must be stored there.
+
+        Every array in the posterior and sample stats groups will have the
+        (chains, tune + draws) batch dimensions to the left of the core dimensions of
+        the model's random variable or the step method's stat shape. The warmup (tuning
+        draws) and the posterior samples are split at a later stage, once
+        :meth:`~ZarrTrace.split_warmup_groups` is called.
+
+        After the creation if the zarr hierarchies, it initializes the list of
+        :class:`~pymc.backends.zarr.Zarrchain` instances (one for each chain) under the
+        ``straces`` attribute. These objects serve as the interface to record draws and
+        samples generated by the step methods for each chain.
+
+        Parameters
+        ----------
+        chains : int
+            The number of chains to use to initialize the arrays.
+        draws : int
+            The number of posterior draws to use to initialize the arrays.
+        tune : int
+            The number of tuning steps to use to initialize the arrays.
+        step : pymc.step_methods.compound.BlockedStep | pymc.step_methods.compound.CompoundStep
+            The step method that will be used to generate the draws and stats.
+        model : pymc.model.core.Model | None
+            If None, the model is taken from the ``with`` context.
+        vars : Sequence[TensorVariable] | None
+            Sampling values will be stored for these variables. If ``None``,
+            ``model.unobserved_RVs`` is used.
+        test_point : dict[str, numpy.ndarray] | None
+            This is not used and is a product of the inheritance of :class:`ZarrChain`
+            from :class:`~.BaseTrace`, which uses it to determine the shape and dtype
+            of `vars`.
+        """
+        if self._is_base_setup:
+            raise RuntimeError("The ZarrTrace has already been initialized")  # pragma: no cover
+        model = modelcontext(model)
+        self.model = model
+        self.coords, self.vars_to_dims = coords_and_dims_for_inferencedata(model)
+        if vars is None:
+            vars = model.unobserved_value_vars
+
+        unnamed_vars = {var for var in vars if var.name is None}
+        assert not unnamed_vars, f"Can't trace unnamed variables: {unnamed_vars}"
+        self.varnames = get_default_varnames(
+            [var.name for var in vars], include_transformed=self.include_transformed
+        )
+        self.vars = [var for var in vars if var.name in self.varnames]
+
+        self.fn = model.compile_fn(self.vars, inputs=model.value_vars, on_unused_input="ignore")
```

Comment:
Allow passing the compiled function already and use `pytensor.In`/`pytensor.Out` and trust_input as was done in https://github.com/pymc-devs/pymc/pull/7578/commits/b589ce8562fe9fb23be420f4208333bd2db8357e

Reply:
Should be addressed by 596a27c. I didn't refactor the precompilation into a helper, but I did add a flag to `model.compile_fn`.